odoo/addons/base/models/ir_mail_server.py

```diff
@@ -181,36 +205,69 @@ def connect(self, host=None, port=None, user=None, password=None, encryption=Non
            :param user: optional username to authenticate with
            :param password: optional password to authenticate with
            :param string encryption: optional, ``'ssl'`` | ``'starttls'``
+           :param smtp_from: FROM SMTP envelop, used to find the best mail server
+           :param ssl_certificate: filename of the SSL certificate used for authentication
+               Used when no mail server is given and overwrite  the odoo-bin argument "smtp_ssl_certificate"
+           :param ssl_private_key: filename of the SSL private key used for authentication
+               Used when no mail server is given and overwrite  the odoo-bin argument "smtp_ssl_private_key"
            :param bool smtp_debug: toggle debugging of SMTP sessions (all i/o
                               will be output in logs)
            :param mail_server_id: ID of specific mail server to use (overrides other parameters)
         """
         # Do not actually connect while running in test mode
-        if getattr(threading.currentThread(), 'testing', False):
-            return None
+        if self._is_test_mode():
+            return
+
+        def load_cert_chain(certificate_filename, private_key_filename):
```

Comment:
Why is this a nested function? It doesn't seem to close over anything, why does it have to be re-created on every `connect` call?

Reply:
I can do it without this nested function if you prefer